### PR TITLE
Performance optimizations

### PR DESCRIPTION
**Pull request by pctablet505**
This pull request introduces a series of optimizations and fast-paths across the Keras backend for PyTorch, targeting performance improvements in tensor operations, attention mechanisms, and global state access. The changes primarily focus on avoiding unnecessary conversions and function calls in common inference scenarios, reducing overhead in frequently used code paths, and improving cache utilization.

Performance optimizations for tensor operations:

* Added fast-paths to functions like `add`, `subtract`, `matmul`, `multiply`, `expand_dims`, and `where` in `keras/src/backend/torch/numpy.py` to avoid unnecessary conversions when inputs are already torch tensors, significantly improving performance during inference. [[1]](diffhunk://#diff-edf2246e48f9f8c0d7db782b20ebc454fb44430626b2873bb1db6f7737191030R68-R70) [[2]](diffhunk://#diff-edf2246e48f9f8c0d7db782b20ebc454fb44430626b2873bb1db6f7737191030R104-R107) [[3]](diffhunk://#diff-edf2246e48f9f8c0d7db782b20ebc454fb44430626b2873bb1db6f7737191030R164-R165) [[4]](diffhunk://#diff-edf2246e48f9f8c0d7db782b20ebc454fb44430626b2873bb1db6f7737191030R810) [[5]](diffhunk://#diff-edf2246e48f9f8c0d7db782b20ebc454fb44430626b2873bb1db6f7737191030L1848-R1871)
* Optimized `convert_to_tensor`, `cast`, and `slice_update` in `keras/src/backend/torch/core.py` to minimize redundant device and dtype conversions, and to use in-place updates where possible, reducing memory overhead and speeding up KV cache updates. [[1]](diffhunk://#diff-8cbf2174ac4e424662c42b4d637778381d815ed0f2993f01677c1a22b15e1e2dR199-R225) [[2]](diffhunk://#diff-8cbf2174ac4e424662c42b4d637778381d815ed0f2993f01677c1a22b15e1e2dR306-R310) [[3]](diffhunk://#diff-8cbf2174ac4e424662c42b4d637778381d815ed0f2993f01677c1a22b15e1e2dL587-R673)
* Improved `while_loop` in `keras/src/backend/torch/core.py` by using a range-based for loop when `maximum_iterations` is known, resulting in faster generation tasks.

Attention and neural network layer optimizations:

* Enhanced `dot_product_attention` and related attention functions in `keras/src/backend/torch/nn.py` by skipping unnecessary tensor conversions and dtype casts when inputs are already torch tensors with matching dtypes, and by optimizing mask and bias handling. [[1]](diffhunk://#diff-d1aa910274af8fda33a1e7f4cc92b54f53ed45261a9f9db56411ae75faa4b006R1155-R1160) [[2]](diffhunk://#diff-d1aa910274af8fda33a1e7f4cc92b54f53ed45261a9f9db56411ae75faa4b006R1172-R1189)
* Added a cache for flash attention availability checks in `_can_use_flash_attention` to avoid repeated import attempts and improve error handling.
* Modified `_compute_attention` in `keras/src/layers/attention/multi_head_attention.py` to use a more efficient fused attention path during inference when dropout is not applied.

Global state and dtype handling improvements:

* Replaced calls to `get_global_attribute` with direct attribute access in global state tracker for faster retrieval of device, name scope, and autocast scope, reducing function call overhead. [[1]](diffhunk://#diff-288889f0a3127ec816d4ff2944bf9dd31d3718a02581ed63b015fc3f5b4fdc75R4-R16) [[2]](diffhunk://#diff-288889f0a3127ec816d4ff2944bf9dd31d3718a02581ed63b015fc3f5b4fdc75L40-R53) [[3]](diffhunk://#diff-288889f0a3127ec816d4ff2944bf9dd31d3718a02581ed63b015fc3f5b4fdc75L58-R77) [[4]](diffhunk://#diff-ef65411710c16fa09f65f31ca1befe257e41339f5aa5bfd77ead35b4df4d364dL667-R684) [[5]](diffhunk://#diff-8cbf2174ac4e424662c42b4d637778381d815ed0f2993f01677c1a22b15e1e2dL74-R78)
* Added a cache for torch dtype conversions in `standardize_dtype` to speed up repeated dtype normalization and reduce mapping overhead. [[1]](diffhunk://#diff-ef65411710c16fa09f65f31ca1befe257e41339f5aa5bfd77ead35b4df4d364dR570-R588) [[2]](diffhunk://#diff-ef65411710c16fa09f65f31ca1befe257e41339f5aa5bfd77ead35b4df4d364dR601-R602)

Symbolic tensor and miscellaneous optimizations:

* Refactored `any_symbolic_tensors` in `keras/src/backend/common/keras_tensor.py` to avoid expensive tree flattening in eager execution, checking only necessary structures for symbolic tensors.
* Added fast-paths to neural network functions like `gelu`, `softmax`, and `dot_product_attention` to skip conversions when inputs are already torch tensors, further optimizing inference workloads. [[1]](diffhunk://#diff-d1aa910274af8fda33a1e7f4cc92b54f53ed45261a9f9db56411ae75faa4b006R122) [[2]](diffhunk://#diff-d1aa910274af8fda33a1e7f4cc92b54f53ed45261a9f9db56411ae75faa4b006R155-R161) [[3]](diffhunk://#diff-d1aa910274af8fda33a1e7f4cc92b54f53ed45261a9f9db56411ae75faa4b006R1155-R1160)

These changes collectively improve the efficiency of Keras with PyTorch backend, especially for inference and generation tasks, by minimizing unnecessary computation and memory usage.

### CODE REVIEW

File: keras/src/backend/torch/numpy.py [hunk 1]
```diff
@@ -65,6 +65,9 @@ def rot90(array, k=1, axes=(0, 1)):
 
 
 def add(x1, x2):
+    # Fast path: both already torch tensors
+    if isinstance(x1, torch.Tensor) and isinstance(x2, torch.Tensor):
```
Comment (jeffcarp) [inline keras/src/backend/torch/numpy.py]: Can this type of check be moved to within `convert_to_tensor` instead of updating all of its callsites?

**Comment by pctablet505** [comment]
@gemini-code-assist review
